File: apps/web/src/lib/gmail-service.ts
/**
 * Gmail Service
 * Handles email sending and receiving using Gmail API with OAuth2 authentication
 */

import { z } from 'zod'

// Email validation schema
export const EmailSchema = z.object({
  to: z.string().email('Invalid email address'),
  subject: z.string().min(1, 'Subject is required'),
  body: z.string().min(1, 'Email body is required'),
  cc: z.string().email().optional(),
  bcc: z.string().email().optional(),
  attachments: z.array(z.object({
    filename: z.string(),
    content: z.string(), // Base64 encoded content
    contentType: z.string()
  })).optional()
})

export type EmailData = z.infer<typeof EmailSchema>

// Gmail credentials schema
export const GmailCredentialsSchema = z.object({
  clientId: z.string(),
  clientSecret: z.string(),
  redirectUri: z.string().url(),
  accessToken: z.string().optional(),
  refreshToken: z.string().optional(),
  expiryDate: z.date().optional()
})

export type GmailCredentials = z.infer<typeof GmailCredentialsSchema>

// Email response schema
export const EmailResponseSchema = z.object({
  success: z.boolean(),
  messageId: z.string().optional(),
  threadId: z.string().optional(),
  error: z.string().optional(),
  provider: z.literal('gmail'),
  timestamp: z.string()
})

export type EmailResponse = z.infer<typeof EmailResponseSchema>

// Gmail message schema
export const GmailMessageSchema = z.object({
  id: z.string(),
  threadId: z.string(),
  snippet: z.string(),
  subject: z.string().optional(),
  from: z.string().optional(),
  to: z.string().optional(),
  date: z.string().optional(),
  body: z.string().optional(),
  attachments: z.array(z.object({
    filename: z.string(),
    size: z.number(),
    attachmentId: z.string()
  })).optional()
})

export type GmailMessage = z.infer<typeof GmailMessageSchema>

/**
 * Gmail service class for sending and receiving emails
 */
export class GmailService {
  private oauth2Client: any
  private gmail: any
  private credentials: GmailCredentials

  constructor(credentials: GmailCredentials) {
    this.credentials = GmailCredentials.parse(credentials)
    this.initializeOAuth2Client()
  }

  /**
   * Initialize OAuth2 client
   */
  private async initializeOAuth2Client() {
    const { google } = await import('googleapis')
    
    this.oauth2Client = new google.auth.OAuth2(
      this.credentials.clientId,
      this.credentials.clientSecret,
      this.credentials.redirectUri
    )

    // Set credentials if available
    if (this.credentials.accessToken) {
      this.oauth2Client.setCredentials({
        access_token: this.credentials.accessToken,
        refresh_token: this.credentials.refreshToken
      })
    }

    // Initialize Gmail API
    this.gmail = google.gmail({ version: 'v1', auth: this.oauth2Client })
  }

  /**
   * Generate OAuth2 authorization URL
   */
  getAuthUrl(scopes: string[] = [
    'https://www.googleapis.com/auth/gmail.send',
    'https://www.googleapis.com/auth/gmail.readonly',
    'https://www.googleapis.com/auth/gmail.modify'
  ]): string {
    return this.oauth2Client.generateAuthUrl({
      access_type: 'offline',
      scope: scopes,
      prompt: 'consent'
    })
  }

  /**
   * Exchange authorization code for tokens
   */
  async exchangeCodeForTokens(code: string): Promise<{
    accessToken: string
    refreshToken: string
    expiryDate: Date
  }> {
    try {
      const { tokens } = await this.oauth2Client.getToken(code)
      
      if (!tokens.access_token || !tokens.refresh_token) {
        throw new Error('Invalid tokens received from Google')
      }

      // Set credentials on client
      this.oauth2Client.setCredentials(tokens)

      // Update credentials
      this.credentials.accessToken = tokens.access_token
      this.credentials.refreshToken = tokens.refresh_token
      this.credentials.expiryDate = tokens.expiry_date 
        ? new Date(tokens.expiry_date) 
        : new Date(Date.now() + 3600 * 1000) // 1 hour from now

      return {
        accessToken: this.credentials.accessToken,
        refreshToken: this.credentials.refreshToken,
        expiryDate: this.credentials.expiryDate
      }
    } catch (error) {
      throw new Error(`Failed to exchange code for tokens: ${error}`)
    }
  }

  /**
   * Refresh access token
   */
  async refreshAccessToken(): Promise<string> {
    if (!this.credentials.refreshToken) {
      throw new Error('No refresh token available')
    }

    try {
      const { credentials } = await this.oauth2Client.refreshAccessToken()
      
      if (!credentials.access_token) {
        throw new Error('Failed to refresh access token')
      }

      this.credentials.accessToken = credentials.access_token
      this.credentials.expiryDate = credentials.expiry_date 
        ? new Date(credentials.expiry_date) 
        : new Date(Date.now() + 3600 * 1000)

      return this.credentials.accessToken
    } catch (error) {
      throw new Error(`Failed to refresh access token: ${error}`)
    }
  }

  /**
   * Ensure valid access token
   */
  private async ensureValidToken(): Promise<void> {
    const now = new Date()
    const isExpired = this.credentials.expiryDate 
      ? this.credentials.expiryDate <= now 
      : false

    if (isExpired || !this.credentials.accessToken) {
      await this.refreshAccessToken()
    }
  }

  /**
   * Send email using Gmail API
   */
  async sendEmail(emailData: EmailData): Promise<EmailResponse> {
    try {
      // Validate email data
      const validatedData = EmailSchema.parse(emailData)
      
      // Ensure we have a valid token
      await this.ensureValidToken()

      // Create email message
      const emailMessage = this.createEmailMessage(validatedData)
      
      // Send email
      const response = await this.gmail.users.messages.send({
        userId: 'me',
        requestBody: {
          raw: emailMessage
        }
      })

      console.log('Email sent successfully:', {
        messageId: response.data.id,
        threadId: response.data.threadId,
        to: validatedData.to,
        subject: validatedData.subject
      })

      return {
        success: true,
        messageId: response.data.id,
        threadId: response.data.threadId,
        provider: 'gmail',
        timestamp: new Date().toISOString()
      }
    } catch (error) {
      console.error('Failed to send email:', error)
      return {
        success: false,
        error: error instanceof Error ? error.message : 'Unknown error',
        provider: 'gmail',
        timestamp: new Date().toISOString()
      }
    }
  }

  /**
   * Create email message in RFC 2822 format
   */
  private createEmailMessage(emailData: EmailData): string {
    const { google } = require('googleapis')
    
    // Create email headers
    const headers = [
      `To: ${emailData.to}`,
      `Subject: ${emailData.subject}`,
      'MIME-Version: 1.0',
      'Content-Type: text/html; charset=utf-8'
    ]

    if (emailData.cc) {
      headers.push(`Cc: ${emailData.cc}`)
    }

    if (emailData.bcc) {
      headers.push(`Bcc: ${emailData.bcc}`)
    }

    // Create message
    const message = [
      ...headers,
      '',
      emailData.body
    ].join('\n')

    // Base64 encode the message
    return Buffer.from(message).toString('base64')
      .replace(/\+/g, '-')
      .replace(/\//g, '_')
      .replace(/=+$/, '')
  }

  /**
   * Get list of emails
   */
  async getEmails(options: {
    maxResults?: number
    query?: string
    pageToken?: string
  } = {}): Promise<{
    messages: GmailMessage[]
    nextPageToken?: string
  }> {
    try {
      await this.ensureValidToken()

      const response = await this.gmail.users.messages.list({
        userId: 'me',
        maxResults: options.maxResults || 10,
        q: options.query || '',
        pageToken: options.pageToken
      })

      if (!response.data.messages) {
        return { messages: [] }
      }

      // Get full message details
      const messages = await Promise.all(
        response.data.messages.map(async (msg: any) => {
          const fullMessage = await this.gmail.users.messages.get({
            userId: 'me',
            id: msg.id,
            format: 'full'
          })

          return this.parseGmailMessage(fullMessage.data)
        })
      )

      return {
        messages,
        nextPageToken: response.data.nextPageToken
      }
    } catch (error) {
      console.error('Failed to get emails:', error)
      throw new Error(`Failed to retrieve emails: ${error}`)
    }
  }

  /**
   * Parse Gmail message data
   */
  private parseGmailMessage(data: any): GmailMessage {
    const headers = data.payload?.headers || []
    
    const getHeader = (name: string) => {
      const header = headers.find((h: any) => h.name.toLowerCase() === name.toLowerCase())
      return header?.value
    }

    // Extract email body
    let body = ''
    if (data.payload?.parts) {
      // Multipart message
      for (const part of data.payload.parts) {
        if (part.mimeType === 'text/html' && part.body?.data) {
          body = Buffer.from(part.body.data, 'base64').toString('utf-8')
          break
        } else if (part.mimeType === 'text/plain' && part.body?.data && !body) {
          body = Buffer.from(part.body.data, 'base64').toString('utf-8')
        }
      }
    } else if (data.payload?.body?.data) {
      // Single part message
      body = Buffer.from(data.payload.body.data, 'base64').toString('utf-8')
    }

    // Extract attachments
    const attachments = data.payload?.parts
      ?.filter((part: any) => part.filename && part.body?.attachmentId)
      ?.map((part: any) => ({
        filename: part.filename,
        size: part.body.size,
        attachmentId: part.body.attachmentId
      })) || []

    return {
      id: data.id,
      threadId: data.threadId,
      snippet: data.snippet || '',
      subject: getHeader('Subject'),
      from: getHeader('From'),
      to: getHeader('To'),
      date: getHeader('Date'),
      body,
      attachments
    }
  }

  /**
   * Get email by ID
   */
  async getEmailById(messageId: string): Promise<GmailMessage> {
    try {
      await this.ensureValidToken()

      const response = await this.gmail.users.messages.get({
        userId: 'me',
        id: messageId,
        format: 'full'
      })

      return this.parseGmailMessage(response.data)
    } catch (error) {
      console.error('Failed to get email:', error)
      throw new Error(`Failed to retrieve email: ${error}`)
    }
  }

  /**
   * Send procurement notification email
   */
  async sendProcurementNotification(
    to: string,
    type: 'quote_received' | 'approval_required' | 'order_placed' | 'delivered',
    data: any
  ): Promise<EmailResponse> {
    const templates = {
      quote_received: {
        subject: 'New Quote Received - SupplyGraph',
        body: `
          <div style="font-family: Arial, sans-serif; max-width: 600px; margin: 0 auto;">
            <h2 style="color: #2563eb;">New Quote Received</h2>
            <p>A new quote has been received for your procurement request.</p>
            <div style="background: #f8fafc; padding: 16px; border-radius: 8px; margin: 16px 0;">
              <h3>Quote Details:</h3>
              <ul style="list-style: none; padding: 0;">
                <li><strong>Vendor:</strong> ${data.vendorName}</li>
                <li><strong>Amount:</strong> $${data.amount}</li>
                <li><strong>Items:</strong> ${data.items?.length || 0}</li>
              </ul>
            </div>
            <p>Please review and approve the quote in your <a href="${process.env.NEXT_PUBLIC_APP_URL}/procurement" style="color: #2563eb;">dashboard</a>.</p>
            <hr style="margin: 24px 0; border: none; border-top: 1px solid #e2e8f0;">
            <p style="color: #64748b; font-size: 14px;">This is an automated message from SupplyGraph.</p>
          </div>
        `
      },
      approval_required: {
        subject: 'Approval Required - SupplyGraph',
        body: `
          <div style="font-family: Arial, sans-serif; max-width: 600px; margin: 0 auto;">
            <h2 style="color: #dc2626;">Approval Required</h2>
            <p>A quote requires your approval before proceeding.</p>
            <div style="background: #fef2f2; padding: 16px; border-radius: 8px; margin: 16px 0; border-left: 4px solid #dc2626;">
              <h3>Quote Details:</h3>
              <ul style="list-style: none; padding: 0;">
                <li><strong>Quote ID:</strong> ${data.quoteId}</li>
                <li><strong>Vendor:</strong> ${data.vendorName}</li>
                <li><strong>Amount:</strong> $${data.amount}</li>
                <li><strong>Requested by:</strong> ${data.requestedBy}</li>
              </ul>
            </div>
            <p>Please review and approve in your <a href="${process.env.NEXT_PUBLIC_APP_URL}/procurement" style="color: #dc2626;">dashboard</a>.</p>
            <hr style="margin: 24px 0; border: none; border-top: 1px solid #e2e8f0;">
            <p style="color: #64748b; font-size: 14px;">This is an automated message from SupplyGraph.</p>
          </div>
        `
      },
      order_placed: {
        subject: 'Order Placed Successfully - SupplyGraph',
        body: `
          <div style="font-family: Arial, sans-serif; max-width: 600px; margin: 0 auto;">
            <h2 style="color: #16a34a;">Order Placed</h2>
            <p>Your order has been successfully placed.</p>
            <div style="background: #f0fdf4; padding: 16px; border-radius: 8px; margin: 16px 0; border-left: 4px solid #16a34a;">
              <h3>Order Details:</h3>
              <ul style="list-style: none; padding: 0;">
                <li><strong>Order ID:</strong> ${data.orderId}</li>
                <li><strong>Vendor:</strong> ${data.vendorName}</li>
                <li><strong>Amount:</strong> $${data.amount}</li>
                <li><strong>Expected Delivery:</strong> ${data.expectedDelivery}</li>
              </ul>
            </div>
            <p>You will receive updates on the delivery status.</p>
            <hr style="margin: 24px 0; border: none; border-top: 1px solid #e2e8f0;">
            <p style="color: #64748b; font-size: 14px;">This is an automated message from SupplyGraph.</p>
          </div>
        `
      },
      delivered: {
        subject: 'Order Delivered - SupplyGraph',
        body: `
          <div style="font-family: Arial, sans-serif; max-width: 600px; margin: 0 auto;">
            <h2 style="color: #0891b2;">Order Delivered</h2>
            <p>Your order has been delivered successfully.</p>
            <div style="background: #f0f9ff; padding: 16px; border-radius: 8px; margin: 16px 0; border-left: 4px solid #0891b2;">
              <h3>Delivery Details:</h3>
              <ul style="list-style: none; padding: 0;">
                <li><strong>Order ID:</strong> ${data.orderId}</li>
                <li><strong>Delivered on:</strong> ${data.deliveredDate}</li>
                <li><strong>Items received:</strong> ${data.itemsReceived}</li>
              </ul>
            </div>
            <p>Please verify the delivery and update inventory.</p>
            <hr style="margin: 24px 0; border: none; border-top: 1px solid #e2e8f0;">
            <p style="color: #64748b; font-size: 14px;">This is an automated message from SupplyGraph.</p>
          </div>
        `
      }
    }

    const template = templates[type]
    if (!template) {
      throw new Error(`Unknown email template type: ${type}`)
    }

    return this.sendEmail({
      to,
      subject: template.subject,
      body: template.body
    })
  }

  /**
   * Get updated credentials
   */
  getCredentials(): GmailCredentials {
    return { ...this.credentials }
  }
}

// Factory function to create Gmail service with stored credentials
export const createGmailService = async (userId: string): Promise<GmailService> => {
  // Import Prisma dynamically to avoid ES module issues
  const { PrismaClient } = await import('@prisma/client')
  const prisma = new PrismaClient()

  try {
    // Get user's Gmail credentials from database
    const userCredentials = await prisma.gmailCredentials.findUnique({
      where: { userId }
    })

    if (!userCredentials) {
      throw new Error('Gmail credentials not found for user')
    }

    return new GmailService({
      clientId: process.env.GOOGLE_CLIENT_ID!,
      clientSecret: process.env.GOOGLE_CLIENT_SECRET!,
      redirectUri: process.env.GOOGLE_REDIRECT_URI!,
      accessToken: userCredentials.accessToken,
      refreshToken: userCredentials.refreshToken,
      expiryDate: userCredentials.expiryDate
    })
  } finally {
    await prisma.$disconnect()
  }
}

export default GmailService

File: apps/ai-service/src/services/gmail_service.py
"""Gmail API service for email integration."""

import base64
import email
from typing import Dict, Any, List, Optional
from datetime import datetime

import structlog
from google.auth.transport.requests import Request
from google.oauth2.credentials import Credentials
from google_auth_oauthlib.flow import Flow
from googleapiclient.discovery import build
from googleapiclient.errors import HttpError

from ..config import get_settings

logger = structlog.get_logger(__name__)


class GmailService:
    """Service for Gmail API integration."""
    
    def __init__(self):
        self.settings = get_settings()
        self.service = None
        self._credentials = None
    
    async def authenticate(self, credentials_data: Optional[Dict[str, Any]] = None) -> str:
        """Authenticate with Gmail API and return authorization URL if needed."""
        if credentials_data:
            self._credentials = Credentials.from_authorized_user_info(credentials_data)
        
        if not self._credentials or not self._credentials.valid:
            if self._credentials and self._credentials.expired and self._credentials.refresh_token:
                self._credentials.refresh(Request())
            else:
                # Need to get new credentials
                flow = Flow.from_client_config(
                    {
                        "web": {
                            "client_id": self.settings.gmail_client_id,
                            "client_secret": self.settings.gmail_client_secret,
                            "auth_uri": "https://accounts.google.com/o/oauth2/auth",
                            "token_uri": "https://oauth2.googleapis.com/token",
                            "redirect_uris": [self.settings.gmail_redirect_uri],
                        }
                    },
                    scopes=["https://www.googleapis.com/auth/gmail.readonly", 
                           "https://www.googleapis.com/auth/gmail.send"],
                )
                flow.redirect_uri = self.settings.gmail_redirect_uri
                
                auth_url, _ = flow.authorization_url(prompt="consent")
                return auth_url
        
        self.service = build("gmail", "v1", credentials=self._credentials)
        return "authenticated"
    
    async def handle_oauth_callback(self, authorization_code: str) -> Dict[str, Any]:
        """Handle OAuth callback and return credentials."""
        flow = Flow.from_client_config(
            {
                "web": {
                    "client_id": self.settings.gmail_client_id,
                    "client_secret": self.settings.gmail_client_secret,
                    "auth_uri": "https://accounts.google.com/o/oauth2/auth",
                    "token_uri": "https://oauth2.googleapis.com/token",
                    "redirect_uris": [self.settings.gmail_redirect_uri],
                }
            },
            scopes=["https://www.googleapis.com/auth/gmail.readonly", 
                   "https://www.googleapis.com/auth/gmail.send"],
        )
        flow.redirect_uri = self.settings.gmail_redirect_uri
        
        flow.fetch_token(code=authorization_code)
        self._credentials = flow.credentials
        self.service = build("gmail", "v1", credentials=self._credentials)
        
        return {
            "token": self._credentials.token,
            "refresh_token": self._credentials.refresh_token,
            "token_uri": self._credentials.token_uri,
            "client_id": self._credentials.client_id,
            "client_secret": self._credentials.client_secret,
            "scopes": self._credentials.scopes,
        }
    
    async def search_emails(self, query: str, max_results: int = 100) -> List[str]:
        """Search for emails matching the query."""
        if not self.service:
            raise ValueError("Gmail service not authenticated")
        
        try:
            result = self.service.users().messages().list(
                userId="me",
                q=query,
                maxResults=max_results
            ).execute()
            
            messages = result.get("messages", [])
            return [msg["id"] for msg in messages]
            
        except HttpError as e:
            logger.error("Failed to search emails", error=str(e))
            raise
    
    async def get_email_details(self, message_id: str) -> Dict[str, Any]:
        """Get detailed information about an email."""
        if not self.service:
            raise ValueError("Gmail service not authenticated")
        
        try:
            message = self.service.users().messages().get(
                userId="me",
                id=message_id,
                format="full"
            ).execute()
            
            return self._parse_email_message(message)
            
        except HttpError as e:
            logger.error("Failed to get email details", message_id=message_id, error=str(e))
            raise
    
    async def send_email(
        self,
        to: str,
        subject: str,
        body: str,
        from_email: Optional[str] = None,
        attachments: Optional[List[Dict[str, Any]]] = None,
    ) -> str:
        """Send an email via Gmail API."""
        if not self.service:
            raise ValueError("Gmail service not authenticated")
        
        try:
            message = self._create_email_message(
                to=to,
                subject=subject,
                body=body,
                from_email=from_email,
                attachments=attachments,
            )
            
            result = self.service.users().messages().send(
                userId="me",
                body=message
            ).execute()
            
            logger.info("Email sent successfully", message_id=result["id"], to=to)
            return result["id"]
            
        except HttpError as e:
            logger.error("Failed to send email", to=to, error=str(e))
            raise
    
    async def send_rfq_email(
        self,
        vendor_email: str,
        vendor_name: str,
        subject: str,
        content: str,
        request_id: str,
    ) -> str:
        """Send an RFQ email to a vendor."""
        # Add request ID to subject for tracking
        full_subject = f"{subject} [REQ:{request_id}]"
        
        # Add footer with request tracking
        full_content = f"""{content}

---
Request ID: {request_id}
This is an automated request from SupplyGraph.
Please reply to this email with your quote.
"""
        
        return await self.send_email(
            to=vendor_email,
            subject=full_subject,
            body=full_content,
        )
    
    def _parse_email_message(self, message: Dict[str, Any]) -> Dict[str, Any]:
        """Parse Gmail API message into standardized format."""
        headers = {h["name"]: h["value"] for h in message["payload"].get("headers", [])}
        
        # Extract basic info
        email_data = {
            "id": message["id"],
            "thread_id": message["threadId"],
            "subject": headers.get("Subject", ""),
            "from": headers.get("From", ""),
            "to": self._parse_recipients(headers.get("To", "")),
            "cc": self._parse_recipients(headers.get("Cc", "")),
            "date": headers.get("Date", ""),
            "received_at": datetime.fromtimestamp(int(message["internalDate"]) / 1000),
            "participants": self._extract_participants(headers),
        }
        
        # Extract body and attachments
        body, attachments = self._extract_body_and_attachments(message["payload"])
        email_data["body"] = body
        email_data["attachments"] = attachments
        
        return email_data
    
    def _parse_recipients(self, recipients_str: str) -> List[str]:
        """Parse recipient string into list of email addresses."""
        if not recipients_str:
            return []
        
        # Simple parsing - can be enhanced for complex formats
        recipients = [r.strip() for r in recipients_str.split(",")]
        return [r for r in recipients if "@" in r]
    
    def _extract_participants(self, headers: Dict[str, str]) -> List[str]:
        """Extract all email participants."""
        participants = set()
        
        for field in ["From", "To", "Cc", "Bcc"]:
            if field in headers:
                participants.update(self._parse_recipients(headers[field]))
        
        return list(participants)
    
    def _extract_body_and_attachments(self, payload: Dict[str, Any]) -> tuple[str, List[Dict[str, Any]]]:
        """Extract email body and attachments from payload."""
        body = ""
        attachments = []
        
        def process_part(part):
            nonlocal body, attachments
            
            mime_type = part.get("mimeType", "")
            
            if mime_type == "text/plain":
                if "data" in part.get("body", {}):
                    body_data = part["body"]["data"]
                    decoded_body = base64.urlsafe_b64decode(body_data).decode("utf-8")
                    body += decoded_body
            
            elif mime_type == "text/html" and not body:
                # Use HTML as fallback if no plain text
                if "data" in part.get("body", {}):
                    body_data = part["body"]["data"]
                    decoded_body = base64.urlsafe_b64decode(body_data).decode("utf-8")
                    # Simple HTML to text conversion
                    import re
                    body = re.sub(r'<[^>]+>', '', decoded_body)
            
            elif mime_type.startswith("multipart/"):
                # Process multipart recursively
                for subpart in part.get("parts", []):
                    process_part(subpart)
            
            else:
                # Check if it's an attachment
                filename = None
                for header in part.get("headers", []):
                    if header["name"] == "Content-Disposition":
                        if "filename=" in header["value"]:
                            filename = header["value"].split("filename=")[1].strip('"')
                            break
                
                if filename or part.get("body", {}).get("attachmentId"):
                    attachment = {
                        "filename": filename or "attachment",
                        "mime_type": mime_type,
                        "size": part.get("body", {}).get("size", 0),
                        "attachment_id": part.get("body", {}).get("attachmentId"),
                    }
                    attachments.append(attachment)
        
        process_part(payload)
        return body.strip(), attachments
    
    def _create_email_message(
        self,
        to: str,
        subject: str,
        body: str,
        from_email: Optional[str] = None,
        attachments: Optional[List[Dict[str, Any]]] = None,
    ) -> Dict[str, Any]:
        """Create email message for Gmail API."""
        import email.mime.text
        import email.mime.multipart
        
        if attachments:
            message = email.mime.multipart.MIMEMultipart()
            message.attach(email.mime.text.MIMEText(body, "plain"))
            
            # TODO: Add attachment support
        else:
            message = email.mime.text.MIMEText(body, "plain")
        
        message["to"] = to
        message["subject"] = subject
        if from_email:
            message["from"] = from_email
        
        raw_message = base64.urlsafe_b64encode(message.as_bytes()).decode("utf-8")
        return {"raw": raw_message}